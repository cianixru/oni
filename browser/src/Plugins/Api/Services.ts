import { getInstance, VersionControlManager } from "./../../Services/VersionControl"

export class Services {
    public get vcs(): VersionControlManager {
        return getInstance()
    }
}
