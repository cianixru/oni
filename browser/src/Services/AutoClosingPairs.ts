/**
 * AutoClosingPairs
 *
 * Service to enable auto-closing pair key bindings
 */

import * as Oni from "oni-api"

import { Configuration } from "./Configuration"
import { EditorManager } from "./EditorManager"
import { InputManager } from "./InputManager"
import { LanguageManager } from "./Language"

import { NeovimInstance } from "./../neovim"

import * as Log from "./../Log"

export interface IAutoClosingPair {
    open: string
    close: string
    // TODO: Support `notIn` equivalent
}

export const activate = (
    configuration: Configuration,
    editorManager: EditorManager,
    inputManager: InputManager,
    languageManager: LanguageManager,
) => {
    const insertModeFilter = () => editorManager.activeEditor.mode === "insert"

    let subscriptions: Oni.DisposeFunction[] = []

    const handleOpenCharacter = (pair: IAutoClosingPair, editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            // TODO: PERFORMANCE: Look at how to collapse this instead of needed multiple asynchronous calls.
            await inputFunc(pair.open + pair.close)

            const pos = await neovim.callFunction("getpos", ["."])
            const [, oneBasedLine, oneBasedColumn] = pos
            await editor.activeBuffer.setCursorPosition(oneBasedLine - 1, oneBasedColumn - 2)
        })

        return true
    }

    const handleBackspaceCharacter = (pairs: IAutoClosingPair[], editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer
            const lines = await activeBuffer.getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
            )
            const line = lines[0]

            const { column } = activeBuffer.cursor

            const matchingPair = pairs.find(p => {
                return column >= 1 && line[column] === p.close && line[column - 1] === p.open
            })

            if (matchingPair) {
                // Remove the pairs
                const beforePair = line.substring(0, column - 1)
                const afterPair = line.substring(column + 1, line.length)

                const pos = await neovim.callFunction("getpos", ["."])
                const [, oneBasedLine, oneBasedColumn] = pos
                await editor.activeBuffer.setCursorPosition(oneBasedLine - 1, oneBasedColumn - 2)

                await activeBuffer.setLines(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.line + 1,
                    [beforePair + afterPair],
                )
            } else {
                await inputFunc("<bs>")
            }
        })

        return true
    }

    const handleEnterCharacter = (pairs: IAutoClosingPair[], editor: Oni.Editor) => () => {
        const neovim: NeovimInstance = editor.neovim as any
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer

            const lines = await (activeBuffer as any).getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
                false,
            )
            const line = lines[0]

            const { column } = activeBuffer.cursor

            const matchingPair = pairs.find(p => {
                return column >= 1 && line[column] === p.close && line[column - 1] === p.open
            })

            if (matchingPair) {
                const whiteSpacePrefix = getWhiteSpacePrefix(line)
                const beforePair = line.substring(0, column)
                const afterPair = line.substring(column, line.length)

                const pos = await neovim.callFunction("getpos", ["."])
                const [, oneBasedLine] = pos
                await activeBuffer.setLines(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.line + 1,
                    [beforePair, whiteSpacePrefix, whiteSpacePrefix + afterPair],
                )
                await activeBuffer.setCursorPosition(oneBasedLine, whiteSpacePrefix.length)
                await inputFunc("<tab>")
            } else {
                await inputFunc("<enter>")
            }
        })

        return true
    }

    const handleCloseCharacter = (pair: IAutoClosingPair, editor: Oni.Editor) => () => {
        const neovim: any = editor.neovim
        neovim.blockInput(async (inputFunc: any) => {
            const activeBuffer = editor.activeBuffer
            const lines = await (activeBuffer as any).getLines(
                activeBuffer.cursor.line,
                activeBuffer.cursor.line + 1,
                false,
            )
            const line = lines[0]
            if (line[activeBuffer.cursor.column] === pair.close) {
                await activeBuffer.setCursorPosition(
                    activeBuffer.cursor.line,
                    activeBuffer.cursor.column + 1,
                )
            } else {
                await inputFunc(pair.close)
            }
        })

        return true
    }

    const onBufferEnter = (newBuffer: Oni.Buffer) => {
        if (!configuration.getValue("autoClosingPairs.enabled")) {
            Log.verbose("[Auto Closing Pairs] Not enabled.")
            return
        }

        if (subscriptions && subscriptions.length) {
            subscriptions.forEach(df => df())
        }

        subscriptions = []

        const autoClosingPairs = getAutoClosingPairs(configuration, newBuffer.language)

        autoClosingPairs.forEach(pair => {
            subscriptions.push(
                inputManager.bind(
                    pair.open,
                    handleOpenCharacter(pair, editorManager.activeEditor),
                    insertModeFilter,
                ),
            )
            subscriptions.push(
                inputManager.bind(
                    pair.close,
                    handleCloseCharacter(pair, editorManager.activeEditor),
                    insertModeFilter,
                ),
            )
        })

        subscriptions.push(
            inputManager.bind(
                "<bs>",
                handleBackspaceCharacter(autoClosingPairs, editorManager.activeEditor),
                insertModeFilter,
            ),
        )
        subscriptions.push(
            inputManager.bind(
                "<enter>",
                handleEnterCharacter(autoClosingPairs, editorManager.activeEditor),
                insertModeFilter,
            ),
        )
    }

    editorManager.activeEditor.onBufferEnter.subscribe(onBufferEnter)

    const activeEditor = editorManager.activeEditor
    if (activeEditor && activeEditor.activeBuffer) {
        onBufferEnter(activeEditor.activeBuffer)
    }
}

const nonWhiteSpaceRegEx = /\S/

export const getWhiteSpacePrefix = (str: string): string => {
    const firstIndex = str.search(nonWhiteSpaceRegEx)

    if (firstIndex === -1) {
        return ""
    } else {
        return str.substring(0, firstIndex)
    }
}

const getAutoClosingPairs = (
    configuration: Configuration,
    language: string,
): IAutoClosingPair[] => {
    const languagePairs = configuration.getValue(`language.${language}.autoClosingPairs`)

    if (languagePairs) {
        return languagePairs
    } else {
        return configuration.getValue("autoClosingPairs.default") || []
    }
}
