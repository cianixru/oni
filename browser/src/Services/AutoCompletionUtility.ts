/**
 * AutoCompletionUtility.ts
 *
 * Helper functions for auto completion
 */

export function getCompletionStart(bufferLine: string, cursorColumn: number, completion: string): number {

    cursorColumn = Math.min(cursorColumn, bufferLine.length)

    let x = cursorColumn
    while (x >= 0) {

        const subWord = bufferLine.substring(x, cursorColumn + 1)

        if (completion.indexOf(subWord) === -1) {
            break
        }

        x--
    }

    return x + 1
}

export function replacePrefixWithCompletion(bufferLine: string, basePosition: number, cursorColumn: number, completion: string): string {
    const startPosition = basePosition

    const before = bufferLine.substring(0, startPosition)
    const after = bufferLine.substring(cursorColumn, bufferLine.length)

    return before + completion + after
}

export interface CompletionMeetResult {
    position: number
    base: string
    shouldExpandCompletions: boolean
}

export const doesCharacterMatchTriggerCharacters = (character: string, triggerCharacters: string[]): boolean => {
    return triggerCharacters.indexOf(character) >= 0
}

/**
 * Returns the start of the 'completion meet' along with the current base for completion
 */
export function getCompletionMeet(line: string, cursorColumn: number, characterMatchRegex: RegExp, completionTriggerCharacters: string[]): CompletionMeetResult {

    // Clamp column to within string bands
    let col = Math.max(cursorColumn - 1, 0)
    col = Math.min(col, line.length - 1)

    let currentPrefix = ""

    while (col >= 0 && col < line.length) {
        const currentCharacter = line[col]

        if (!currentCharacter.match(characterMatchRegex) || doesCharacterMatchTriggerCharacters(currentCharacter, completionTriggerCharacters)) {
            break
        }

        currentPrefix = currentCharacter + currentPrefix
        col--
    }

    const basePos = col

    const isFromTriggerCharacter = doesCharacterMatchTriggerCharacters(line[basePos], completionTriggerCharacters)

    const shouldExpandCompletions = currentPrefix.length > 0 || isFromTriggerCharacter

    // If the expansion is due to a letter, start the match at the letter position
    const position = isFromTriggerCharacter ? basePos : basePos + 1

    return {
        position: position + 1,
        base: currentPrefix,
        shouldExpandCompletions,
    }
}
