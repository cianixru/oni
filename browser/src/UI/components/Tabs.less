@import (reference) "./common.less";

@keyframes tab-icon-appear {
    0% {
        transform: scale(0.5);
    }
    50% {
        transform: scale(1.25);
    }
    100% {
        transform: scale(1);
    }
}

.tab-icon-appear-animation {
    animation-name: tab-icon-appear;
    animation-duration: 0.6s;
    animation-function: ease-in;
    animation-fill-mode: forwards;
    opacity: 1;
}

.tabs {
    display: flex;
    flex-direction: row;
    align-items: flex-end;

    width: 100%;
    overflow-x: hidden;

    transform: translateY(-3px);
    transition: transform 0.25s ease;

    .loaded & {
        transform: translateY(0px);
    }

    &:hover {
        overflow-x: overlay;
    }
}

.tabs::-webkit-scrollbar {
    height: 3px;
}

.tab {
    display: flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    cursor: pointer;

    flex: 0 0 auto;

    background-color: rgb(40, 44, 52);
    opacity: 0.6;
    transition: opacity 0.25s;

    overflow: hidden;
    user-select: none;

    .name {
        flex: 1 1 auto;
        text-align: center;
        height: 100%;

        display: flex;
        justify-content: center;
        align-items: center;
        margin: 0px 4px;
    }

    .icon-container {
        position: absolute;
        top: 0px;
        left: 0px;
        right: 0px;
        bottom: 0px;

        display: flex;
        justify-content: center;
        align-items: center;

        transition: opacity 0.25s ease-out;
    }

    .x-icon-container {
        opacity: 0;
    }

    .circle-icon-container {
        opacity: 0;
    }

    &.is-dirty {
        .circle-icon-container {
            .tab-icon-appear-animation;
        }
    }

    &.not-dirty:hover {
        .corner {
            .x-icon-container {
                .tab-icon-appear-animation;
            }
        }
    }

    .corner {
        flex: 0 0 auto;
        width: 32px;
        height: 100%;

        display: flex;
        align-items: center;
        justify-content: center;

        position: relative;

        .circle {
            width: 8px;
            height: 8px;
            border-radius: 4px;
            background-color: #c8c8c8;
        }
    }

    .corner.enable-hover {
        &:hover {
            background-color: rgba(100, 100, 100, 0.1);
        }
    }

    &.not-selected {
        .box-shadow-up-inset();

        &:hover {
            opacity: 0.9;
        }
    }

    &.selected {
        .box-shadow-up();
        opacity: 1;
    }
}
