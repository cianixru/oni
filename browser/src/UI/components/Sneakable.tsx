/**
 * Sneakable
 *
 * Helper for easily integrating 'sneak mode' with UI elements
 */

import * as React from "react"

import { Shapes } from "oni-api"
import { IDisposable } from "oni-types"

import { /* SneakProvider, Sneak,*/ getInstance as getSneak } from "./../../Services/Sneak"

export interface ISneakableProps {
    callback?: () => void
}

export class Sneakable extends React.PureComponent<ISneakableProps, {}> {
    private _subscription: IDisposable
    private _element: HTMLDivElement = null

    public componentDidMount() {
        this._cleanupSubscription()

        this._subscription = getSneak().addSneakProvider(() => {
            if (this._element) {
                const rect = this._element.getBoundingClientRect()

                return [
                    {
                        callback: this.props.callback
                            ? this.props.callback
                            : () => this._element.click(),
                        rectangle: Shapes.Rectangle.create(
                            rect.left,
                            rect.top,
                            rect.width,
                            rect.height,
                        ),
                    },
                ]
            } else {
                return null
            }
        })
    }

    public componentWillUnmount(): void {
        this._cleanupSubscription()
    }

    public _cleanupSubscription(): void {
        if (this._subscription) {
            this._subscription.dispose()
            this._subscription = null
        }
    }

    public render(): JSX.Element {
        return <div ref={elem => (this._element = elem)}>{this.props.children}</div>
    }
}
