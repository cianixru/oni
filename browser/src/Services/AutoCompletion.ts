/**
 * AutoCompletion.ts
 *
 * Manages the autocompletion service:
 *  - Integrating with Neovim's external popup menu
 *  - Managing completion state
 */

import * as types from "vscode-languageserver-types"

import { IBuffer, INeovimInstance } from "./../neovim"

import * as Utility from "./AutoCompletionUtility"

import * as UI from "./../UI/index"

export class AutoCompletion {

    constructor(
        private _neovimInstance: INeovimInstance,
    ) {
        this._neovimInstance.on("show-popup-menu", (completions: any[]) => {
            const c = completions.map((completion) => ({
                kind: types.CompletionItemKind.Text,
                label: completion[0],
            }))

            UI.Actions.showCompletions({
                base: "",
                completions: c,
            })
        })

        this._neovimInstance.on("hide-popup-menu", () => {
            UI.Actions.hideCompletions()
        })
    }

    public complete(): void {
        let cursorRow: number
        let cursorColumn: number
        let originalLineLength: number
        let newLineLength: number

        const completion = UI.Selectors.getSelectedCompletion() || ""
        let currentBuffer: IBuffer
        this._neovimInstance.getCurrentBuffer()
            .then((buffer) => currentBuffer = buffer)

            // Get position in buffer
            .then(() => this._neovimInstance.getCursorColumn())
            .then((col) => cursorColumn = col)

            .then(() => this._neovimInstance.getCursorRow())
            .then((row) => cursorRow = row)

            // Get current line
            .then(() => {
                return currentBuffer.getLines(cursorRow - 1, cursorRow, false)
            })
            // Replace with completion value
            .then((value) => {
                const line = value[0]
                originalLineLength = line.length
                const newLine = Utility.replacePrefixWithCompletion(line, cursorColumn - 2, completion)
                newLineLength = newLine.length
                return currentBuffer.setLines(cursorRow - 1, cursorRow, false, [newLine])
            })
            .then(() => {
                const cursorOffset = newLineLength - originalLineLength

                // Set cursor position after the word
                return this._neovimInstance.eval(`setpos(".", [0, ${cursorRow}, ${cursorColumn + cursorOffset}, 0])`)
            })

        UI.Actions.hideCompletions()
    }
}
