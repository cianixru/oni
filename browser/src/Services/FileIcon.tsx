/**
 * Icons
 *
 * - Data source for icons present in Oni
 * - Loads icons based on the `ui.iconTheme` configuration setting
 */

import * as React from "react"

import { css, keyframes, styled, withProps } from "../UI/components/common"
import { getInstance } from "./IconThemes"

const appearAnimationKeyframes = keyframes`
    0% {
        opacity: 0;
        transform: scale(0.8);
    }
    100% {
        opacity: 1;
        transform: scale(1);
    }
`

const appearAnimation = css`
    animation-name: ${appearAnimationKeyframes};
    animation-duration: 0.25s;
    animation-timing-function: ease-in;
    animation-fill-mode: forwards;
    opacity: 1;
`

const Icon = withProps<{ playAppearAnimation: boolean }>(styled.i)`
    ${props => (props.playAppearAnimation ? appearAnimation : "")}
`

interface IFileIconProps {
    fileName: string
    language?: string

    isLarge?: boolean

    playAppearAnimation?: boolean
}

export const FileIcon = (props: IFileIconProps) => {
    if (!props.fileName) {
        return null
    }

    const icons = getInstance()

    const className =
        icons.getIconClassForFile(props.fileName, props.language) + (props.isLarge ? " fa-lg" : "")

    return (
        <Icon
            playAppearAnimation={props.playAppearAnimation}
            className={className}
            aria-hidden={true}
        />
    )
}

export const getFileIcon = (fileName: string) => <FileIcon fileName={fileName} isLarge={true} />
