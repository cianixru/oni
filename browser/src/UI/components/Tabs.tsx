/**
 * Tabs.tsx
 */

import * as path from "path"

import * as React from "react"
import { connect } from "react-redux"

import * as classNames from "classnames"

import * as BufferSelectors from "./../../Editor/NeovimEditor/NeovimEditorSelectors"
import * as State from "./../../Editor/NeovimEditor/NeovimEditorStore"

import { addDefaultUnitIfNeeded } from "./../../Font"

import { Icon } from "./../../UI/Icon"

import { FileIcon } from "./../../Services/FileIcon"

export interface ITabProps {
    id: number
    name: string
    description: string
    isSelected: boolean
    isDirty: boolean
    iconFileName?: string
    highlightColor?: string
}

export interface ITabContainerProps {
    onBufferSelect?: (bufferId: number) => void
    onBufferClose?: (bufferId: number) => void

    onTabSelect?: (tabId: number) => void
    onTabClose?: (tabId: number) => void
}

export interface ITabsProps {
    onSelect?: (id: number) => void
    onClose?: (id: number) => void

    visible: boolean
    tabs: ITabProps[]

    backgroundColor: string
    foregroundColor: string

    shouldWrap: boolean
    maxWidth: string
    height: string

    fontFamily: string
    fontSize: string
}

export class Tabs extends React.PureComponent<ITabsProps, {}> {
    public render(): JSX.Element {
        if (!this.props.visible) {
            return null
        }

        const wrapStyle: React.CSSProperties = {
            flexWrap: "wrap",
        }

        const overflowStyle = this.props.shouldWrap ? wrapStyle : {}

        const tabBorderStyle: React.CSSProperties = {
            ...overflowStyle,
            borderBottom: `4px solid ${this.props.backgroundColor}`,
            fontFamily: this.props.fontFamily,
            fontSize: this.props.fontSize,
        }

        const tabs = this.props.tabs.map(t => {
            return (
                <Tab
                    key={t.id}
                    {...t}
                    onClickName={() => this._onSelect(t.id)}
                    onClickClose={() => this._onClickClose(t.id)}
                    backgroundColor={this.props.backgroundColor}
                    foregroundColor={this.props.foregroundColor}
                    height={this.props.height}
                    maxWidth={this.props.maxWidth}
                />
            )
        })

        return (
            <div className="tabs horizontal enable-mouse layer" style={tabBorderStyle}>
                {tabs}
            </div>
        )
    }

    private _onSelect(id: number): void {
        this.props.onSelect(id)
    }

    private _onClickClose(id: number): void {
        this.props.onClose(id)
    }
}

export interface ITabPropsWithClick extends ITabProps {
    onClickName: React.EventHandler<React.MouseEvent<HTMLDivElement>>
    onClickClose: React.EventHandler<React.MouseEvent<HTMLDivElement>>

    backgroundColor: string
    foregroundColor: string

    height: string
    maxWidth: string
}

export const Tab = (props: ITabPropsWithClick) => {
    const cssClasses = classNames("tab", {
        selected: props.isSelected,
        "not-selected": !props.isSelected,
        "is-dirty": props.isDirty,
        "not-dirty": !props.isDirty,
    })

    const style = {
        backgroundColor: props.backgroundColor,
        color: props.foregroundColor,
        maxWidth: props.maxWidth,
        height: props.height,
        borderTop: "2px solid " + props.highlightColor,
    }

    return (
        <div className={cssClasses} title={props.description} style={style}>
            <div className="corner" onClick={props.onClickName}>
                <FileIcon
                    fileName={props.iconFileName}
                    isLarge={true}
                    additionalClassNames={"file-icon-appear-animation"}
                />
            </div>
            <div className="name" onClick={props.onClickName}>
                <span className="name-inner">{props.name}</span>
            </div>
            <div className="corner enable-hover" onClick={props.onClickClose}>
                <div className="icon-container x-icon-container">
                    <Icon name="times" />
                </div>
                <div className="icon-container circle-icon-container">
                    <div className="circle" />
                </div>
            </div>
        </div>
    )
}

const getTabName = (name: string): string => {
    if (!name) {
        return "[No Name]"
    }

    return path.basename(name)
}

import { createSelector } from "reselect"

const getTabState = (state: State.IState) => state.tabState

const sanitizedModeForColors = (mode: string): string => {
    if (mode === "showmatch") {
        return "insert"
    }

    return mode
}

export const getHighlightColor = (state: State.IState) => {
    if (!state.configuration["tabs.highlight"] || !state.hasFocus) {
        return "transparent"
    }

    const sanitizedMode = sanitizedModeForColors(state.mode)
    const colorForMode = "highlight.mode." + sanitizedMode + ".background"
    const color = state.colors[colorForMode]
    return color || "transparent"
}

export const showTabId = (state: State.IState) => {
    return state.configuration["tabs.showIndex"]
}

export const getIdPrefix = (id: string, shouldShow: boolean): string => {
    return shouldShow ? id + ": " : ""
}

export const shouldShowFileIcon = (state: State.IState): boolean => {
    return state.configuration["tabs.showFileIcon"]
}

const getTabsFromBuffers = createSelector(
    [
        BufferSelectors.getBufferMetadata,
        BufferSelectors.getActiveBufferId,
        getHighlightColor,
        showTabId,
        shouldShowFileIcon,
    ],
    (
        allBuffers: any,
        activeBufferId: any,
        color: string,
        shouldShowId: boolean,
        showFileIcon: boolean,
    ) => {
        const bufferCount = allBuffers.length
        const tabs = allBuffers.map((buf: any): ITabProps => {
            const isActive =
                (activeBufferId !== null && buf.id === activeBufferId) || bufferCount === 1
            return {
                id: buf.id,
                name: getIdPrefix(buf.id, shouldShowId) + getTabName(buf.file),
                iconFileName: showFileIcon ? getTabName(buf.file) : "",
                highlightColor: isActive ? color : "transparent",
                isSelected: isActive,
                isDirty: buf.modified,
                description: buf.file,
            }
        })
        return tabs.sort(({ id: prevId }: ITabProps, { id: nextId }: ITabProps) => prevId - nextId)
    },
)

const getTabsFromVimTabs = createSelector(
    [getTabState, getHighlightColor, showTabId, shouldShowFileIcon],
    (tabState: any, color: any, shouldShowId: boolean, showFileIcon: boolean) => {
        return tabState.tabs.map((t: any, idx: number) => ({
            id: t.id,
            name: getIdPrefix((idx + 1).toString(), shouldShowId) + getTabName(t.name),
            highlightColor: t.id === tabState.selectedTabId ? color : "transparent",
            iconFileName: showFileIcon ? getTabName(t.name) : "",
            isSelected: t.id === tabState.selectedTabId,
            isDirty: false,
            description: t.name,
        }))
    },
)

const mapStateToProps = (state: State.IState, ownProps: ITabContainerProps): ITabsProps => {
    const oniTabMode = state.configuration["tabs.mode"]
    const shouldUseVimTabs = oniTabMode === "tabs"

    const tabs = shouldUseVimTabs ? getTabsFromVimTabs(state) : getTabsFromBuffers(state)

    const visible = oniTabMode !== "native" && oniTabMode !== "hidden"

    const height = state.configuration["tabs.height"]
    const maxWidth = state.configuration["tabs.maxWidth"]
    const shouldWrap = state.configuration["tabs.wrap"]

    const selectFunc = shouldUseVimTabs ? ownProps.onTabSelect : ownProps.onBufferSelect
    const closeFunc = shouldUseVimTabs ? ownProps.onTabClose : ownProps.onBufferClose

    return {
        fontFamily: state.configuration["ui.fontFamily"],
        fontSize: addDefaultUnitIfNeeded(state.configuration["ui.fontSize"]),
        backgroundColor: state.colors["tabs.background"],
        foregroundColor: state.colors["tabs.foreground"],
        onSelect: selectFunc,
        onClose: closeFunc,
        height,
        maxWidth,
        shouldWrap,
        visible,
        tabs,
    }
}

export const TabsContainer = connect(mapStateToProps)(Tabs)
