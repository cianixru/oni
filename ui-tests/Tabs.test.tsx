import { mount, shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"
import * as React from "react"

jest.mock("../browser/src/Services/FileIcon")
import { FileIcon } from "../browser/src/Services/FileIcon"
jest.mock("../browser/src/UI/components/Sneakable")
import { Sneakable } from "../browser/src/UI/components/Sneakable"

import { Corner, getHighlightColor, Name, Tab, Tabs } from "../browser/src/UI/components/Tabs"

const MockFileIcon = FileIcon as jest.Mock<{}>
const MockSneakable = Sneakable as jest.Mock<Sneakable>

describe("<Tabs /> Tests", () => {
    MockFileIcon.mockImplementation(() => ({ render: jest.fn() }))
    MockSneakable.mockImplementation(props => ({
        render: () => props.children,
    }))

    const testTheme: any = {
        "highlight.mode.normal.background": "green",
        "highlight.mode.insert.background": "blue",
    }

    const tabCloseFunction = jest.fn()
    const tabSelectFunction = jest.fn()

    const tab1 = {
        id: 1,
        name: "test",
        description: "a test tab",
        isSelected: true,
        isDirty: true,
        iconFileName: "icon",
        highlightColor: "#000",
    }

    const tab2 = {
        id: 2,
        name: "test",
        description: "a test tab",
        isSelected: false,
        isDirty: true,
        iconFileName: "icon",
        highlightColor: "#000",
    }

    const TabsContainingSingleTab = (
        <Tabs
            fontSize="1.2em"
            shouldShowHighlight
            mode="normal"
            maxWidth="20em"
            height="2em"
            fontFamily="inherit"
            shouldWrap={false}
            visible={true}
            onTabClose={tabCloseFunction}
            onTabSelect={tabSelectFunction}
            tabs={[tab1]}
        />
    )

    const TabsContainingTwoTabs = (
        <Tabs
            fontSize="1.2em"
            shouldShowHighlight
            mode="normal"
            maxWidth="20em"
            height="2em"
            fontFamily="inherit"
            shouldWrap={false}
            visible={true}
            onTabClose={tabCloseFunction}
            onTabSelect={tabSelectFunction}
            tabs={[tab1, tab2]}
        />
    )

    const TabsNotVisible = (
        <Tabs
            fontSize="1.2em"
            shouldShowHighlight
            mode="normal"
            maxWidth="20em"
            height="2em"
            fontFamily="inherit"
            shouldWrap={false}
            visible={false}
            tabs={[tab1]}
        />
    )

    const unfocusedTab = (
        <Tabs
            fontSize="1.2em"
            shouldShowHighlight={false}
            mode="normal"
            maxWidth="20em"
            height="2em"
            fontFamily="inherit"
            shouldWrap={false}
            visible={false}
            tabs={[tab1]}
        />
    )

    afterEach(() => {
        tabCloseFunction.mockReset()
        tabSelectFunction.mockReset()
    })

    it("renders without crashing", () => {
        const wrapper = shallow(TabsContainingSingleTab)
        expect(wrapper.length).toEqual(1)
    })

    it("should match last known snapshot unless we make a change", () => {
        const wrapper = shallow(TabsContainingSingleTab)
        expect(shallowToJson(wrapper)).toMatchSnapshot()
    })

    it("should render the correct number of tabs", () => {
        expect(shallow(TabsContainingSingleTab).children().length).toEqual(1)
        expect(shallow(TabsContainingTwoTabs).children().length).toEqual(2)
    })

    it("should not render if the visible prop is false", () => {
        const wrapper = shallow(TabsNotVisible)
        expect(wrapper.getElement()).toBe(null)
    })

    it("should call onTabClose callback on tab close button click", () => {
        const wrapper = mount(TabsContainingSingleTab)
        const clickedTab = wrapper.find(Tab)

        wrapper
            .find(Corner)
            .last()
            .simulate("click")

        expect(tabCloseFunction).toHaveBeenCalledWith(clickedTab.props().id)

        wrapper.unmount()
    })

    it("should call onTabSelect callback on tab title click", () => {
        const wrapper = mount(TabsContainingTwoTabs)
        const clickedTab = wrapper.find(Tab).last()

        clickedTab.find(Name).simulate("mouseDown", { button: 0 })
        expect(tabSelectFunction).toHaveBeenCalledWith(clickedTab.props().id)

        wrapper.unmount()
    })

    it("should call onTabClose callback on tab title middle click", () => {
        const wrapper = mount(TabsContainingTwoTabs)
        const clickedTab = wrapper.find(Tab).first()

        clickedTab.find(Name).simulate("mouseDown", { button: 1 })
        expect(tabCloseFunction).toHaveBeenCalledWith(clickedTab.props().id)

        wrapper.unmount()
    })

    it("should call onTabSelect callback on tab file icon click", () => {
        const wrapper = mount(TabsContainingTwoTabs)
        const clickedTab = wrapper.find(Tab).last()

        clickedTab
            .find(Corner)
            .first()
            .simulate("mouseDown", { button: 0 })
        expect(tabSelectFunction).toHaveBeenCalledWith(clickedTab.props().id)

        wrapper.unmount()
    })

    it("should call onTabClose callback on tab file icon middle click", () => {
        const wrapper = mount(TabsContainingTwoTabs)
        const clickedTab = wrapper.find(Tab).first()

        clickedTab
            .find(Corner)
            .first()
            .simulate("mouseDown", { button: 1 })
        expect(tabCloseFunction).toHaveBeenCalledWith(clickedTab.props().id)

        wrapper.unmount()
    })

    it("should pass tab name as Sneakable tag property", () => {
        const wrapper = mount(TabsContainingSingleTab)
        const tab = wrapper.find(Tab)
        const sneakable = wrapper.find(Sneakable)

        expect(sneakable.props().tag).toEqual(tab.props().name)

        wrapper.unmount()
    })

    it("should call onTabSelect callback as Sneakable callback", () => {
        MockSneakable.mockImplementationOnce(props => {
            props.callback()
        })

        const wrapper = mount(TabsContainingSingleTab)
        const sneakableTab = wrapper.find(Tab)

        expect(tabSelectFunction).toHaveBeenCalledWith(sneakableTab.props().id)

        wrapper.unmount()
    })

    it("should pass tab icon file name as FileIcon fileName property ", () => {
        const wrapper = mount(TabsContainingSingleTab)
        const tab = wrapper.find(Tab)
        const fileIcon = wrapper.find(FileIcon)

        expect(fileIcon.props().fileName).toEqual(tab.props().iconFileName)

        wrapper.unmount()
    })

    it("should have a transparent border if not selected", () => {
        const highlight = getHighlightColor({
            theme: testTheme,
            isSelected: false,
            shouldShowHighlight: true,
            mode: "normal",
        })
        expect(highlight).toBe("transparent")
    })

    it("should highlight the active tab with the mode color", () => {
        const highlight = getHighlightColor({
            theme: testTheme,
            isSelected: true,
            shouldShowHighlight: true,
            mode: "normal",
        })
        expect(highlight).toBe("green")
    })

    it("should highlight the active tab with the normal mode color if the mode is a commandline mode", () => {
        const highlight = getHighlightColor({
            theme: testTheme,
            isSelected: true,
            shouldShowHighlight: true,
            mode: "cmdline_normal",
        })

        expect(highlight).toBe("green")
    })

    it("should highlight the active tab with the normal mode color if the mode is in showmatch mode", () => {
        const highlight = getHighlightColor({
            theme: testTheme,
            isSelected: true,
            shouldShowHighlight: true,
            mode: "showmatch",
        })

        expect(highlight).toBe("blue")
    })
})
