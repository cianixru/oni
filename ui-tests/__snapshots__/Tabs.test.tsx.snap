// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> Tests should match last known snapshot unless we make a change 1`] = `
<div
  className="tabs horizontal enable-mouse layer"
  style={
    Object {
      "borderBottom": "4px solid #fff",
      "fontFamily": "inherit",
      "fontSize": "1.2em",
    }
  }
>
  <Tab
    backgroundColor="#fff"
    description="a test tab"
    foregroundColor="#000"
    height="2em"
    highlightColor="#000"
    iconFileName="icon"
    id={1}
    isDirty={true}
    isSelected={true}
    key="1"
    maxWidth="20em"
    name="test"
    onClose={[MockFunction]}
    onSelect={[MockFunction]}
  />
</div>
`;
